### PR TITLE
Reassessing incremental clustering engine

### PR DESCRIPTION
**Pull request by jgriss**
@ruiwanguk This pull request tracks the development of the new ReassessingIncrementalClusteringEngine.

Features still missing:
- [ ] Update Factory: discuss how to deal with the fact the ReassessingIncrementalClusteringEngine does require a retainSimilarityThreshold

### CODE REVIEW

**Comment by jgriss** [comment]
This new ClusteringEngine has a major performance issue: in the original code and also the "default" _ClusteringEngine_ class we first cluster all spectra, then merge the clusters, and then we check for non-fitting spectra within the clusters. In this implementation, the checking for non-fitting spectra and the merging are done whenever a new spectrum is being processed.
Checking for non-fitting is not as bad since only "dirty" clusters are being checked. But merging is currently being triggered for all clusters. This needs to be fixed somehow.
- [ ] Optimize merging procedure

File: src/main/java/uk/ac/ebi/pride/spectracluster/engine/ReassessingIncrementalClusteringEngine.java [hunk 1]
```diff
@@ -0,0 +1,432 @@
...
+                mostSimilarCluster = cluster;
+            }
+        }
+
+        if (mostSimilarCluster != null) {
+            // add to cluster
+            ISpectrum[] clusteredSpectra = new ISpectrum[clusteredSpectraToAdd.size()];
+            final ISpectrum[] merged = clusteredSpectraToAdd.toArray(clusteredSpectra);
+            mostSimilarCluster.addSpectra(merged);
+            mostSimilarCluster.setProperty(CLUSTER_PROPERTY_IS_DIRTY, "");
```
Comment (ruiwanguk) [inline src/main/java/uk/ac/ebi/pride/spectracluster/engine/ReassessingIncrementalClusteringEngine.java]: Good idea on setting the cluster as dirty, however, I am not sure using empty string is the best practice here. If it is a boolean, I would recommend to use boolean.

File: src/main/java/uk/ac/ebi/pride/spectracluster/engine/ReassessingIncrementalClusteringEngine.java [hunk 2]
```diff
@@ -0,0 +1,432 @@
...
+        List<ICluster> clustersToremove = findClustersTooLow(precursorMz);
+        // either add as an existing cluster if make a new cluster
+        addToClusters(added);
+
+        // remove any non-fitting spectra
+        List<ICluster> nonFittingSpectra = removeNoneFittingSpectra();
+
+        // try to add them again by merging all clusters
+        internalGetClusters().addAll(nonFittingSpectra);
+        mergeAllClusters();
```
Comment (ruiwanguk) [inline src/main/java/uk/ac/ebi/pride/spectracluster/engine/ReassessingIncrementalClusteringEngine.java]: The logic here should probably be arranged in a different order. mergeAllClusters will produce dirty clusters. When the next time addClusterINcremental is called, findClustersTooLow is called first, which will emit clustersToRemove that could contain dirty clusters as well.

**Comment by ruiwanguk** [comment]
About the retain threshold, in fact, it is used by the Hadoop reducers when writing out the clustering results. If you take a look at [AbstractClusterReducer](https://github.com/spectra-cluster/spectra-cluster-hadoop/blob/master/src/main/java/uk/ac/ebi/pride/spectracluster/hadoop/util/AbstractClusterReducer.java#L70), you should see that the retain threshold is used.

**Comment by jgriss** [comment]
@ruiwanguk I completely missed this feature! Also since I simply didn't expect it to be part of the Hadoop code. In that case, this class is not needed. I therefore suggest we simply delete the branch and close this pull request.
I'd also suggest to add a _**DemergingClusteringEngine**_ to the spectra-cluster API that only removes non-fitting spectra and that works similar to the IncrementalClusteringEngine. This ClusteringEngine could then be used by external tools to simplify the creation of a clustering pipeline, _ie_:
1. IncrementalClusteringEngine (initial clustering)
2. DemergingClusteringEngine (de-merging)
3. IncrementalClusteringEngine (1st merging step)
  ...
If you agree I'd open an issue to keep track of this proposed feature.

**Comment by ruiwanguk** [comment]
@jgriss I am not sure I understand fully the purpose of _DemergingClusteringEngine_.  Does it do incremental clustering? If I get your idea right, you want to use the retain threshold to de-merge the clusters before writing them out. If this is the case, I think we can extend the _IncrementalClusteringEngine_ directly. 

Do correct me if I got your idea wrong.

**Comment by jgriss** [comment]
This feature is not needed. Instead IncrementalClusteringEngine will be updated to support de-merging of spectra (see Issue #5).

**Pull request closed without merging**